### DISCUSSION TITLE
macOS open_config (and so Cmd+,) behavior is obscure

### DISCUSSION THREAD

**Discussion by donaldguy** [ghostty-org/ghostty#3282] [Feature Requests, Ideas] [answered]
while not technically incorrect, documentation is perhaps misleading with the amount of details omitted when it says

```
$ ghostty +list-actions --docs | grep -A3 open_config
open_config:
  Open the configuration file in the default OS editor. If your default OS
  editor isn't configured then this will fail. Currently, any failures to
  open the configuration will show up only in the logs.
```

I've spelled it out at length - but the short of it is mainly: it ends up being treated as a completely unknown file type ([UTI](https://en.wikipedia.org/wiki/Uniform_Type_Identifier) of `public.data`) rather than even `public.plain-text`, so even if a user has put previous effort into defining a macOS-wide "default [text] editor", the result of getting TextEdit (or possibly something associated with some other binary format file previously) is worse-than-expected. 

there are several ways this could be improved

## what happens currently:

As best as I've been able to put it together:

### 1. `open` is shelled out to
per

https://github.com/ghostty-org/ghostty/blob/35b9ceee2116331b83c0c86269394e2545070b0f/src/os/open.zig#L9-L15

Ghostty (after checking for file existence elsewhere) invokes [`open`](https://ss64.com/mac/open.html) `<the config file>`

It is worth noting that `file://<absolute path>` behaves the same and may be something it gets translated to as a matter of course ([unlike](https://developer.apple.com/library/archive/technotes/tn2022/_index.html) prior to 2011?) - since open/LaunchServices really would like to speak urls

### 2. xattr consideration

`open` (as well as finder, so maybe really  launchservicesd?) checks for extended attribute `com.apple.LaunchServices.OpenWith` 
    
if present this is read and parsed as a plist containing a top level dict with keys `version`, `path` and `bundleIdentifier` 
    
- `path` appears to be used preferentially, so long as it is valid absolute path of an `.app`
- if `path` absent or non-existent, `bundleIdentifier` is used instead
- if `bundleIdentifier` also does not exist and specify a valid bundle id (or if the attribute value is not a valid plist, the whole xattr is silently ignored)

this is what is set when using option-click "Open With > Other" and the "Always Open With" checkbox.
Done ^that way Finder writes a binary plist with version 0 and redundant matching `bundleIdentifier` and `path`, but it appears (after testing with `xattr -w`) an xml format plist will also be respected. (but not output of `plutil -convert json`)

### 3. UTI determination 

Failing that, macOS determines a UTI: https://developer.apple.com/documentation/uniformtypeidentifiers
 
specifically LaunchServices does, unclear if in the form of `launchservicesd` 

(The determination may be short circuited by the kMDItemContentType metadata attribute and thus actually be done only at spotlight indexing time? I could not personally with xattrs or content get a different answer doing `mdimport -d 3 -t` on the file; but I imagine a proper spotlight importer extension might change the answer

regardless of when it actually happens:)

Broadly it appears to be a best-fit "conformance" check based on registered equivalences

its kind of documented in a few places:
1.  https://developer.apple.com/documentation/coreservices/launch_services) with a lot of the API surface there is deprecated without a clear pointer at preferred alternative.
2. The old https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/understanding_utis/understand_utis.tasks/understand_utis_tasks.html
3. Linking back into modern docs from there: https://developer.apple.com/documentation/uniformtypeidentifiers/uttype-swift.struct and https://developer.apple.com/documentation/uniformtypeidentifiers/uttypereference

but what happens at the root to decide amongst system defined types remains very unclear to me. There are definitely the "tags":
    
- file extensions and since there is no `.` here that's _probably_ irrelevant
- apparently sometimes mime-type matching, but it is not clear in what circumstances that happens [apparently coming from clipboard? but not apparently (currently) on filesystem.
    - for the record `/usr/bin/file --mime-type ~/.config/ghostty/config` gives `text/plain` which is what I'd expect but is misleading)
 - ?? It then goes to the kernel level. there are mentions of "OSType" as the third tag type, which does seem to be the same (not helpfully) as this: https://developer.apple.com/documentation/kernel/ostype - but couldn't find more helpful docs and [searching the source](https://github.com/search?q=repo%3Aapple-oss-distributions%2Fxnu%20OSType&type=code) just muddied the waters further.

I am not sure whether with no file extension, the fate of the config file is pre-ordained, or if indeed it could entirely depend on what programs and libraries have been installed on the mac in question

(see also: output of `/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister -dump` on your machine.  

There are also notions of Intents and entitlements in there - but possibly limited to for apple originating types? [where this database is kept on disk is also unclear. certainly not a plist]

Mine for example, until I uninstalled (and once I reinstalled) VSCode, had this entry
```
claim id:                   Configuration file (0x2e010)
localizedNames:             "af" = ?, "am" = ?, "ar" = ?, "bg" = ?, "bn" = ?, "ca" = ?, "cs" = ?, "da" = ?, "de" = ?, "el" = ?, "en" = ?, "en_GB" = ?, "es" = ?, "es_419" = ?, "et" = ?, "fa" = ?, "fi" = ?, "fil" = ?, "fr" = ?, "gu" = ?, "he" = ?, "hi" = ?, "hr" = ?, "hu" = ?, "id" = ?, "it" = ?, "ja" = ?, "kn" = ?, "ko" = ?, "LSDefaultLocalizedValue" = "Configuration file", "lt" = ?, "lv" = ?, "ml" = ?, "mr" = ?, "ms" = ?, "nb" = ?, "nl" = ?, "pl" = ?, "pt_BR" = ?, "pt_PT" = ?, "ro" = ?, "ru" = ?, "sk" = ?, "sl" = ?, "sr" = ?, "sv" = ?, "sw" = ?, "ta" = ?, "te" = ?, "th" = ?, "tr" = ?, "uk" = ?, "ur" = ?, "vi" = ?, "zh_CN" = ?, "zh_TW" = ?
rank:                       Default
bundle:                     Code (0x11e80)
flags:                      legacy-wildcard  doc-type  relative-icon-path (0000000000001030)
roles:                      Editor (0000000000000004)
iconFiles:                  Contents/Resources/config.icns
bindings:                   .config, .editorconfig, .ini, .cfg, 'TEXT', 'utxt', 'TUTX', '****'
```

I thought this might in fact be relevant (because e.g. I wasn't actually sure if `^config$` matched a "file extension" of `config` (when I first, for example, ran [`duti`](https://github.com/moretension/duti/)`-x config`). I'm pretty sure no though? )

### 4. (Fallback) Treatment as fully generic `public.data`
   
in the end _on my machine_, UTI conformance decision falls all the way through to "well I know its bytes!" and comes back with fully generic `public.data`. 

In this it has failed to decide on any of the more specific seemingly applicable system types: 
- `public.delimited-values-text`, 
- perhaps `public.source-code`,
- `public.text`, 
- `public.utf8-plain-text`, 
- and, indeed despite what the mime-type would drive if it were actually relevant:
    
#### importantly, that means it isn't `public.plain-text`

This was my source of most confusion, because this is frankly what I would have assumed. 

There is also here the fact that `open` can take a `-t` flag, which man page describes as:

```
     -t  Causes the file to be opened with the default text editor, as determined via LaunchServices
```

and `open --help` slightly more concisely as

```
      -t                    Opens with default text editor.
```

Looking back at what ghostty says of the action: I would expect this to be what "in the default OS editor." equates to

but at this point in my earliest effort to control what Ghostty was doing for Cmd+, I found here that:

`open -t ~/.config/ghostty/config` did what I wanted/expected, while `open ~/.config/ghostty/config` was still "wrong"

## what might be better:

1. minimally: changing `open.zig` to use the `-t` flag for this so you at least get whats configured for plain text files (e.g. associated for `.txt`) [also what I tried to do with `defaults write` after first searching "macos default editor" today] (provided open isn't being used elsewhere that is incompatible with that)
2. do something more OS native instead of going back to Zig at 
    https://github.com/ghostty-org/ghostty/blob/35b9ceee2116331b83c0c86269394e2545070b0f/macos/Sources/Ghostty/Ghostty.App.swift#L139-L142
    then shelling out.
    
    - I think Apple would "want" you to use [`NSWorskpace.open(_:configuration:completionHandler:)`](https://developer.apple.com/documentation/appkit/nsworkspace/open(_:configuration:completionhandler:)). I don't see anything on `OpenConfiguration` adequate to do anything different, but I might be overlooking
     - but I think you can do "better" with [`LSOpenFromURLSpec`](https://developer.apple.com/documentation/coreservices/1441986-lsopenfromurlspec) - at least with `passThruParams` or possibly by doing a more pro-active lookup of another type (by tag, or possibly pointing directly at something like those VSCode "configuration file" registrations) and use that to specify app

3. inform the system of the specific file type (using [LSRegisterURL](https://developer.apple.com/documentation/coreservices/1446350-lsregisterurl), that `lsregister` cli, or–if indeed its possible to do without e.g. adding a file extension (and you definitely don't want to do that)–the packaging [way](https://developer.apple.com/documentation/uniformtypeidentifiers/defining-file-and-data-types-for-your-app) of something like:

    ```xml
    <key>UTExportedTypeDeclarations</key>
    <array>
        <dict>
            <key>UTTypeIdentifier</key>
            <string>com.mitchellh.ghostty-configuration</string>
            <key>UTTypeConformsTo</key>
            <array>
                <string>public.source-code</string>
            </array>
            <key>UTTypeDescription</key>
            <string>Configuration file for Ghostty</string>
            <key>UTTypeTagSpecification</key>
            <dict>
                ... this part remains unclear to me ... 
            </dict>
        </dict>
    </array>
    ```

    in `Info.plist`. 

    Without any further user interventions before or after, that's likely to "just" shunt the default from TextEdit to XCode if its been installed. But if, in fact, a user has previously set up associations to point at another preferred editor, those would be applied)?

    There may be a better equivalence to choose as well (certainly if toml, ini, or java properties were around, I think those would be ~right-ish, but they are not apparently core types [despite several other `com.sun.java-` types being in there. `public.delimited-values-text` seems also correct, but Idk if that has some risk (despite csv and tsv being more specific and in there) of ending up going to e.g. Numbers.app )

    There may also be things doable with e.g. `UTTypeReferenceURL` that I don't understand still. (though the one instance of reference URL I see in my lsregister output is rfc4180.txt for `public.comma-separated-values-text`, soooo maybe thats just `https://ghostty.org/docs/config/reference` (not sure if it is surfaced anywhere)

4. Building on above (or e.g. "cheating" and using the xattr), you could have macOS treat Ghostty itself as the assigned editor and then you could (possibly prompt on first instance? or just have a default [that can try EDITOR, VISUAL env vars?], but for override) have a `macos-config-editor` config (that would have been loaded on last config reload, so shouldn't hit a bootstrapping problem when Ghostty is asked to open the file. That value then just goes to literally or programatic equivalent of the `-a` or `-b` flags on `open` )

    technically you could also just do that config for handling the open_config action (or the Cmd-, case specifically) - but it seems better and in the spirit of Ghostty's doing OS native-ness to also put out the association system wide (it might then more easily be consumable by e.g. editor support extensions if that happens in future)

**Comment by mitchellh** [comment]
Thank you for the detailed writeup. I think this is better left as a discussion which we then peel off specific actionable issues. This sort of long formed discussion is _exactly_ what discussions are for in this repository. 😄 

I'll note that the plan is ultimately to:

1. Default to a native settings GUI: #2354 
2. Have a button to fallback to loading in your `$EDITOR`.

This should address suggestion 2 and 4 you proposed.

But in the short term if we can get some quick wins to improve this then I'd be happy to take it. I think the `open -t` point is really good and we should go with that at the very least to start.

**Comment by mitchellh** [comment]
#3284 for suggestion 1.

**Comment by kuro337** [comment]
Adding to this - I'm mostly running into similar doubts, but maybe from a different perspective, would appreciate any info/thoughts!

I ran into this rabbithole because a super common thing I do is - open the scrollback in $EDITOR

From the settings it says - 

```bash
write_scrollback_file
Write the entire scrollback into a temporary file. The action determines what to do with the filepath. Valid values are:

"paste": Paste the file path into the terminal.
"open": Open the file in the default OS editor for text files. The default OS editor is determined by using open on macOS and xdg-open on Linux.
```

So this opens it in the default TextEdit app, it can be overridden for example - but does -e currently only work with Linux?


I think the docs could be improved too, for example it doesn't explicitly mention anywhere that passing args to ghostty only works in Linux (not sure, correct me if I'm wrong, but I assumed so since it throws this for any -e attempt)

```
   if (comptime build_config.app_runtime == .none) {
        const stdout = std.io.getStdOut().writer();
        try stdout.print("Usage: ghostty +<action> [flags]\n\n", .{});
        try stdout.print(
            \\This is the Ghostty helper CLI that accompanies the graphical Ghostty app.
            \\To launch the terminal directly, please launch the graphical app
            \\(i.e. Ghostty.app on macOS). This CLI can be used to perform various
            \\actions such as inspecting the version, listing fonts, etc.
            \\
            \\We don't have proper help output yet, sorry! Please refer to the
            \\source code or Discord community for help for now. We'll fix this in time.
            \\
        ,
            .{},
        );

        posix.exit(0);
    }
```



So if the ghostty CLI doesn't accept args, then I figured we can pass args to the App bundle?

It seems like after trying out a bunch of these commands, I kept getting similar errors - 

```
open -a Ghostty.app --args -e vim file.txt
# any variation of args - seems to give the same errors
```

So it still wasn't clear to me if it's supported to be able to pass args to ghostty - didn't want to spend too much time on it;
I figured this would be relevant to the discussion as I can't be the only person looking for more relevant information - (at the very least by sharing some perspective from a new user, and common use cases) 

@donaldguy  I see you mentioned this

but at this point in my earliest effort to control what Ghostty was doing for Cmd+, I found here that:

```
open -t ~/.config/ghostty/config did what I wanted/expected, while open ~/.config/ghostty/config was still "wrong"
```

I am assuming this is a patch - or is it something user specified for example while opening the Application?

**Reply by donaldguy** [reply]
I'm not sure, but I think you were asking how to set the default text editor?

If you google it, there are some `defaults write`  invocations targeting LSHandlers that might work (but might not cover role correctly), but probably what you want is, either:

1. to just use a tool like [`duti`](https://github.com/moretension/duti), or [SwiftDefaultApps](https://github.com/Lord-Kamina/SwiftDefaultApps) (which offers both a prefsPane and the `swda` cli); both are in homebrew. You want to target the `public.plain-text` UTI (and `public.data` if your concern is the ghostty config). Where those cli tools offer use of a file extension, `.txt` is probs equivalent (though technically it's tagged across e.g. the encoding specific UTIs as well)

2. Go through a conventional GUI flow around opening a file with extension `.txt` (the Open With > Other and checkbox might be sufficient? Like I mentioned there is that xattr that might set it file specific, but that might actually have just been the behavior used by Finder for an extensionless file like the ghostty config)

Honestly, Apple could/should do better at surfacing this (alongside default browser and email), but perhaps they assume its a concern past a technical threshold (but its odd there is not at least an official CLI. There is `lsapps` which communicates with the same LaunchServices subsystem, but it doesn't have file association relevant subcommands that I've figured out)

**Comment by Integralist** [comment]
I hate to add a useless comment onto a discussion thread, but as I wasn't able to adequately follow along, I feel I should still comment for those users (like me) who aren't _that_ technically advanced but generally know what they would like the end result to be and can just emoji 👍🏻 if this is what they expect/want...

In essence, I have `$EDITOR` set to `nvim` and so I'd like `cmd+,` to open the config in Neovim rather than TextEdit.

Feel free to delete this comment if this requirement is already understood and is something y'all have taken into consideration and don't need explicitly called out in a separate comment.

Thanks again to everyone who works on making ghostty happen. I appreciate you all ❤️

**Reply by edrozenberg** [reply]
Same here, I can't follow the deep details on this. On my Mac Ghostty opens the `config` file in one of my apps (Code Runner, a kind of IDE), which is definitely not what I want. There is no easy way to change this behavior, tried messing w/mime plist (`~/Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure.plist`) for plain-text etc.

Ideally either should use the `EDITOR` setting as you suggest or optionally have a setting in `config` for what editor to use to open the config file. Since the `config` file has no file extension there's no easy obvious solution. Personally I'd want to be able to configure it to use TextMate or vim or MacVim.

**Reply by 00-kat** [reply]
I'm not sure if this is relevant to the current thread.  However, what you requested is closely related to #2504.

**Comment by mitchellh** [comment, marked as answer]
I'm going to close this. We've since moved to launch services with an explicit text filetype, and longer term we're going to move to a preference panel. I appreciate the detail in this suggestion, it led to some tangible improvements in the mean time.
